### PR TITLE
Instance discovery during authority check

### PR DESCRIPTION
**Pull request by trwalke**
Fixes # .
https://github.com/AzureAD/microsoft-authentication-library-for-dotnet/issues/2736

**Changes proposed in this request**
Calling instance discovery during authority check for aliases to be used when validating the authority environments. If a change in authority is detected but the change uses one of the aliased authorites, MSAL should not throw the authority mismatch error

**Testing**
Unit test

**Performance impact**
Evaluating

### CODE REVIEW

File: src/client/Microsoft.Identity.Client/AppConfig/ApplicationConfiguration.cs [hunk 1]
```diff
@@ -134,6 +134,11 @@ public string ClientVersion
         public ITokenCacheInternal AppTokenCacheInternalForTest { get; set; }
 
         public IDeviceAuthManager DeviceAuthManagerForTest { get; set; }
+
+        public void SetAuthorityInfoForTest(AuthorityInfo authorityInfo)
+        {
+            AuthorityInfo = authorityInfo;
```
Comment (bgavrilMS) [inline src/client/Microsoft.Identity.Client/AppConfig/ApplicationConfiguration.cs:140]: Why is this needed? AuthorityInfo already has a setter at line 68.

File: src/client/Microsoft.Identity.Client/ClientApplicationBase.cs [hunk 2]
```diff
@@ -81,8 +81,8 @@ internal ClientApplicationBase(ApplicationConfiguration config)
             RequestContext requestContext,
             ITokenCacheInternal cache)
         {
-            var authority = Microsoft.Identity.Client.Instance.Authority.CreateAuthorityForRequest(
-               requestContext.ServiceBundle.Config.AuthorityInfo,
+            var authority = Instance.Authority.CreateAuthorityForRequest(
+               requestContext,
```
Comment (bgavrilMS) [inline src/client/Microsoft.Identity.Client/ClientApplicationBase.cs:85]: It's better to send exactly the objects needed (in this case the InstanceDiscoveryManager) rather than sending the entire  context object.

File: src/client/Microsoft.Identity.Client/Instance/Authority.cs [hunk 3]
```diff
@@ -208,6 +218,22 @@ private static void ValidateSameHost(AuthorityInfo requestAuthorityInfo, Authori
             }
         }
 
+        private static bool IsAuthorityAliased(RequestContext requestContext, AuthorityInfo requestAuthorityInfo)
+        {
+            var instanceDiscoveryManager = requestContext.ServiceBundle.InstanceDiscoveryManager;
+            var result = instanceDiscoveryManager.GetMetadataEntryAsync(requestContext.ServiceBundle.Config.AuthorityInfo, requestContext).Result;
```
Comment (bgavrilMS) [inline src/client/Microsoft.Identity.Client/Instance/Authority.cs:224]: We should not do this. This will block the thread. Never use `.Result` in production code. This is an async call, so you must make all the methods in the call chain async. There aren't that many.

File: src/client/Microsoft.Identity.Client/Instance/Authority.cs [hunk 4]
```diff
@@ -208,6 +218,22 @@ private static void ValidateSameHost(AuthorityInfo requestAuthorityInfo, Authori
             }
         }
 
+        private static bool IsAuthorityAliased(RequestContext requestContext, AuthorityInfo requestAuthorityInfo)
```
Comment (bgavrilMS) [inline src/client/Microsoft.Identity.Client/Instance/Authority.cs:221]: This verification is expensive, as it involves making a network call. We should only do it if the Config authority and the Request authority are different.

File: tests/Microsoft.Identity.Test.Integration.netfx/HeadlessTests/UsernamePasswordIntegrationTests.cs [hunk 5]
```diff
@@ -1,290 +1,290 @@
...
+using Microsoft.Identity.Test.LabInfrastructure;
+using Microsoft.Identity.Test.Unit;
+using Microsoft.VisualStudio.TestTools.UnitTesting;
+
+namespace Microsoft.Identity.Test.Integration.HeadlessTests
+{
+    // Note: these tests require permission to a KeyVault Microsoft account;
+    // Please ignore them if you are not a Microsoft FTE, they will run as part of the CI build
+    [TestClass]
+    public class UsernamePasswordIntegrationTests
```
Comment (bgavrilMS) [inline tests/Microsoft.Identity.Test.Integration.netfx/HeadlessTests/UsernamePasswordIntegrationTests.cs:29]: Please check your space settings. They should be the default.

File: tests/Microsoft.Identity.Test.Integration.netfx/HeadlessTests/UsernamePasswordIntegrationTests.cs [hunk 6]
```diff
@@ -1,290 +1,290 @@
...
+using Microsoft.Identity.Client.TelemetryCore;
+using Microsoft.Identity.Client.Utils;
+using Microsoft.Identity.Test.Common;
+using Microsoft.Identity.Test.Common.Core.Helpers;
+using Microsoft.Identity.Test.Integration.net45.Infrastructure;
+using Microsoft.Identity.Test.LabInfrastructure;
+using Microsoft.Identity.Test.Unit;
+using Microsoft.VisualStudio.TestTools.UnitTesting;
+
+namespace Microsoft.Identity.Test.Integration.HeadlessTests
```
Comment (bgavrilMS) [inline tests/Microsoft.Identity.Test.Integration.netfx/HeadlessTests/UsernamePasswordIntegrationTests.cs:24]: We should not add an integration test for this. Unit tests are sufficient.

File: tests/Microsoft.Identity.Test.Unit/ApiConfigTests/AuthorityTests.cs [hunk 7]
```diff
@@ -1,15 +1,19 @@
...
+using Microsoft.Identity.Test.Common;
 using Microsoft.Identity.Test.Common.Core.Helpers;
+using Microsoft.Identity.Test.Common.Core.Mocks;
 using Microsoft.VisualStudio.TestTools.UnitTesting;
 
 namespace Microsoft.Identity.Test.Unit.ApiConfigTests
 {
     [TestClass]
-    public class AuthorityTests
+    public class AuthorityTests : TestBase
```
Comment (bgavrilMS) [inline tests/Microsoft.Identity.Test.Unit/ApiConfigTests/AuthorityTests.cs:16]: We should have a unit test that describes this issue.

**Review by bgavrilMS — CHANGES_REQUESTED** [review]
We can't block the thread while making network calls.

File: src/client/Microsoft.Identity.Client/AppConfig/ApplicationConfiguration.cs (hunk 1 above)
Comment (trwalke) [reply src/client/Microsoft.Identity.Client/AppConfig/ApplicationConfiguration.cs:140]: I think I added it because intellisense was telling me that it was not settable. I think it was probably just a bug. Should have verified

File: src/client/Microsoft.Identity.Client/ClientApplicationBase.cs (hunk 2 above)
Comment (trwalke) [reply src/client/Microsoft.Identity.Client/ClientApplicationBase.cs:85]: The instance discovery methods need the request context and since the instance discovery manager and the authority info is in the request context. I think it is better to just send the one object alone. This also reduces the number of parameters that need to be passed around.

https://github.com/AzureAD/microsoft-authentication-library-for-dotnet/blob/4a0d46b37fb067163df88262ab3db473a81609c7/src/client/Microsoft.Identity.Client/Instance/Authority.cs#L224

File: src/client/Microsoft.Identity.Client/Instance/Authority.cs (hunk 3 above)
Comment (trwalke) [reply src/client/Microsoft.Identity.Client/Instance/Authority.cs:224]: makes sense

File: src/client/Microsoft.Identity.Client/Instance/Authority.cs [hunk 8]
```diff
@@ -208,6 +218,22 @@ private static void ValidateSameHost(AuthorityInfo requestAuthorityInfo, Authori
...
+
+            foreach (string alias in result.Aliases)
+            {
+                if (string.Equals(alias, requestAuthorityInfo.Host))
+                {
+                    return true;
+                }
+            }
+
+            return false;
```
Comment (pmaytak) [inline src/client/Microsoft.Identity.Client/Instance/Authority.cs:234]:
```suggestion
            return result.Aliases.Any(alias => alias.Equals(requestAuthorityInfo.Host));
```

**Review by pmaytak — APPROVED** [review]
LGTM so far, pending making that method call async.

File: tests/Microsoft.Identity.Test.Integration.netfx/HeadlessTests/UsernamePasswordIntegrationTests.cs (hunk 6 above)
Comment (trwalke) [reply tests/Microsoft.Identity.Test.Integration.netfx/HeadlessTests/UsernamePasswordIntegrationTests.cs:24]: I did not add a test here. this file was changed by VS for line endings

File: tests/Microsoft.Identity.Test.Integration.netfx/HeadlessTests/UsernamePasswordIntegrationTests.cs (hunk 5 above)
Comment (trwalke) [reply tests/Microsoft.Identity.Test.Integration.netfx/HeadlessTests/UsernamePasswordIntegrationTests.cs:29]: Thsi came from a reccomended change to line endings by VS. not sure why but it complained about it

File: src/client/Microsoft.Identity.Client/Instance/Authority.cs (hunk 8 above)
Comment (trwalke) [reply src/client/Microsoft.Identity.Client/Instance/Authority.cs]: Gotta remember to use Any().
Thanks

File: tests/Microsoft.Identity.Test.Unit/ApiConfigTests/AuthorityTests.cs [hunk 9]
```diff
@@ -33,75 +54,107 @@ public void VerifyAuthorityTest()
...
-
-            var ex3 = Assert.ThrowsException<MsalClientException>(
-             () => Authority.CreateAuthorityForRequest(
-                 AuthorityInfo.FromAdfsAuthority(TestConstants.ADFSAuthority, true),
+            _harness.HttpManager.AddInstanceDiscoveryMockHandler();
+            _testRequestContext.ServiceBundle.Config.HttpManager = _harness.HttpManager;
+            _testRequestContext.ServiceBundle.Config.AuthorityInfo = s_commonAuthority;
+            var ex = Assert.ThrowsException<AggregateException>(
+                () =>
+                Authority.CreateAuthorityForRequestAsync(_testRequestContext, s_ppeAuthority, null).Result);
```
Comment (trwalke) [inline tests/Microsoft.Identity.Test.Unit/ApiConfigTests/AuthorityTests.cs:117]: Now that this is async the test ended up failing. This was the easiest way I found to assert the original conditions when using async in a lambda method. When I tried using the await Authority.CreateAuthorityForRequestAsync() the exception was not thrown and when using .Result the exception is aggregate so I just decided to go with that.

Is there a better way?

**Review by bgavrilMS — APPROVED** [review]
